R/tar.R

```diff
@@ -0,0 +1,215 @@
+  }
+}
+
+#' Check if we need to use R's internal tar implementation
+#'
+#' This is slow, because we need to start an R child process, and the
+#' implementation is also very slow. So it is better to use an extranl tar
+#' program, if we can. We test this by trying to uncompress a .tar.gz
+#' archive using the external program. The name of the tar program is
+#' taken from the `TAR` environment variable, if this is unset then `tar`
+#' is used.
+#'
+#' @return Whether we need to use the internal tar implementation.
+#' @keywords internal
+
+need_internal_tar <- (function() {
+  internal <- NULL
+  function() {
+    if (!is.null(internal)) return(internal)
+    mkdirp(tmp <- tempfile())
+    on.exit(unlink(tmp, recursive = TRUE), add = TRUE)
+    tarfile <- system.file(package = .packageName, "tools", "pkg_1.0.0.tgz")
+    p <- external_untar_process$new(tarfile, exdir = tmp)
+    p$wait(timeout = 2000)
+    p$kill()
+    internal <<- p$get_exit_status() != 0 ||
+      !file.exists(file.path(tmp, "pkg", "DESCRIPTION"))
+    internal
+  }
+})()
+
+#' @importFrom R6 R6Class
+
+external_untar_process <- R6Class(
+  "external_untar_process",
+  inherit = callr::process,
+
+  public = list(
+    initialize = function(tarfile, files = NULL, exdir = ".",
+                          restore_times = TRUE,
+                          tar = Sys.getenv("TAR", "tar"),
+                          post_process = NULL)
+      eup_init(self, private, super, tarfile, files, exdir,
+               restore_times, tar, post_process)
+  ),
+
+  private = list(
+    options = NULL
+  )
+)
+
+r_untar_process <- R6Class(
+  "r_untar_process",
+  inherit = callr::r_process,
+
+  public = list(
+    initialize = function(tarfile, files = NULL, exdir = ".",
+                          restore_times = TRUE, post_process = NULL)
+      runtar_init(self, private, super, tarfile, files, exdir,
+                  restore_times, tar, post_process)
+  ),
+
+  private = list(
+    options = NULL
+  )
+)
+
+eup_init <- function(self, private, super, tarfile, files, exdir,
+                     restore_times, tar, post_process) {
+
+  private$options <- list(
+    tarfile = normalizePath(tarfile),
+    files = files,
+    exdir = exdir,
+    restore_times = restore_times,
+    tar = tar)
+
+  private$options$args <- eup_get_args(private$options)
+  super$initialize(tar, private$options$args, post_process = post_process,
+                   stdout = "|", stderr = "|")
+  invisible(self)
+}
+
+eup_get_args <- function(options) {
+  c(
+    "-x", "-f", options$tarfile,
+    "-C", options$exdir,
+    get_untar_decompress_arg(options$tarfile),
+    if (! options$restore_times) "-m",
+    options$files
+  )
+}
+
+get_untar_decompress_arg <- function(tarfile) {
+  type <- detect_package_archive_type(tarfile)
+  switch(
+    type,
+    "gzip" = "-z",
+    "bzip2" = "-j",
+    "xz" = "-J",
+    "zip" = stop("Not a tar file, looks like a zip file"),
+    "unknown" = character()
```

Comment:
I think most recent versions of tar will automatically detect the compression type and it does not need to be explicitly specified. e.g. from https://www.gnu.org/software/tar/manual/html_section/tar_68.html#SEC135

> The only case when you have to specify a decompression option while reading the archive is when reading from a pipe or from a tape drive that does not support random access

And from https://www.freebsd.org/cgi/man.cgi?query=bsdtar&sektion=1

> In extract or	list modes, this option	is ignored.  Note that,	unlike
> 	     other tar implementations,	this implementation recognizes bzip2
>	     compression automatically when reading archives.

But maybe this is not the case for the tar in RTools?

Reply:
The detection is probably very cheap, we just read the first 6 bytes, so it could be worth being on the safe side.  I am sure that there are a number of tar implementations out there.

Reply:
That is true, I was more concerned with it being extra code we need to maintain (albeit pretty straightforward and should not change often (if ever)).

Reply:
These "fingerprints" are widely used,  so even if they extend the format,  I don't they would . change them. So I am not too concerned about that.